File: src/main/scala/com/vc/easy/L783.scala
package com.vc.easy

class TreeNode783(var _value: Int) {
  var value: Int = _value
  var left: TreeNode783 = null
  var right: TreeNode783 = null
}

object L783 {
  def minDiffInBST(root: TreeNode783): Int = {
    import scala.collection.mutable
    val st = new mutable.Stack[TreeNode783]()
    var current = root
    var res = Int.MaxValue
    var prev = -1
    while(current != null || st.nonEmpty) {
      if(current != null) {
        st.push(current)
        current = current.right
      }
      else {
        current = st.pop
        if(prev != -1) {
          res = Math.min(res, prev - current.value)
        }
        prev = current.value
        current = current.left
      }
    }
    res
  }
}

File: src/main/java/com/vc/easy/L783.java
package com.vc.easy;

import com.vc.hard.TreeNode;

import java.util.Stack;

class L783 {
    public int minDiffInBST(TreeNode root) {
        TreeNode current = root;
        Stack<TreeNode> st =new Stack<>();
        TreeNode prev = null;
        int minDiff = Integer.MAX_VALUE;
        while(current != null || !st.isEmpty()) {
            if(current != null) {
                st.push(current);
                current = current.left;
            }
            else {
                current = st.pop();
                if(prev != null) {
                    minDiff = Math.min(minDiff, current.val - prev.val);
                }
                prev = current;
                current = current.right;
            }
        }
        return minDiff;
    }
}
